include/kernels/staggered_kd_apply_xinv_kernel.cuh

```diff
@@ -1,219 +1,110 @@
 #include <index_helper.cuh>
 #include <register_traits.h>
-#include <block_reduce_helper.h>
-#include <shared_memory_cache_helper.cuh>
+#include <kernel.h>
 
 namespace quda {
 
-  template <typename vFloatSpinor_, typename vFloatGauge_, int coarseDof_, int fineColor_,
-            bool dagger_, typename fineColorSpinor, typename xInvGauge>
-  struct ApplyStaggeredKDBlockArg : kernel_param<> {
+  template <typename Float, int nColor_, int dagger_>
+  struct StaggeredKDBlockArg : kernel_param<> {
+    using real = typename mapper<Float>::type;
 
-    using vFloatSpinor = vFloatSpinor_;
-    using vFloatGauge = vFloatGauge_;
+    static constexpr int nDim = 4;
 
-    using real = typename mapper<vFloatGauge>::type;
+    static constexpr int nSpin = 1;
+    static constexpr int nColor = nColor_;
+    static constexpr bool spin_project = false;
+    static constexpr bool spinor_direct_load = false; // seems to be legacy, copied from dslash_staggered.cuh
+    using F = typename colorspinor_mapper<Float, nSpin, nColor, spin_project, spinor_direct_load>::type;
 
-    static constexpr int fineColor = fineColor_;
-    static constexpr int fineSpin = 1;
-    static constexpr int coarseDof = coarseDof_;
-    static constexpr bool dagger = dagger_;
-
-    static constexpr int blockSizeKD = 16; /** Elements per KD block */
-    static constexpr int paddedSpinorSizeKD = blockSizeKD + 1; // padding is currently broken + 1; /** Padded size */
-
-    static constexpr int xinvRowTileSize = 16; /** Length of a tile for KD */
-    static constexpr int xinvColTileSize = 16; /** Length of a tile for KD */
-    static constexpr int xinvPaddedColTileSize = xinvColTileSize + 1; /** Padding to avoid bank conflicts */
-    static constexpr int numTiles = (coarseDof + xinvColTileSize - 1) / xinvColTileSize; /** Number of tiles, should always be three. */
+    static constexpr QudaReconstructType reconstruct = QUDA_RECONSTRUCT_NO;
+    static constexpr bool gauge_direct_load = false; // seems to be legacy, copied from dslash_staggered.cuh
+    using X = typename gauge_mapper<Float, reconstruct, 18, QUDA_STAGGERED_PHASE_NO, gauge_direct_load, QUDA_GHOST_EXCHANGE_PAD>::type;
 
-    fineColorSpinor out;      /** Output staggered spinor field */
-    const fineColorSpinor in; /** Input staggered spinor field */
-    const xInvGauge xInv;     /** Kahler-Dirac inverse field */
-
-    int_fastdiv x_size[QUDA_MAX_DIM];   /** Dimensions of fine grid */
-    int_fastdiv xc_size[QUDA_MAX_DIM];  /** Dimensions of coarse grid */
+    static constexpr bool dagger = dagger_;
 
-    const int fineVolumeCB;
-    const int_fastdiv coarseVolumeCB;   /** Coarse grid volume */
+    F out;               // output vector field
+    const F in;          // input vector field
+    const X xInv;        // Kahler-Dirac inverse gauge field
+    int_fastdiv X0h;     // One-half of X dimension length
+    int_fastdiv dim[4];  // full lattice dimensions
+    const int volumeCB;  // checkerboarded volume
```

Comment:
All these variable comments should include doxygen markup

Reply:
https://github.com/lattice/quda/commit/28e91cb86b4f40d405a75eefdacb4348d03437eb